tests/page-wait-for-response.spec.ts

```diff
@@ -63,6 +63,44 @@ it('should work with predicate', async ({page, server}) => {
   expect(response.url()).toBe(server.PREFIX + '/digits/2.png');
 });
 
+it('should work with async predicate', async ({page, server}) => {
+  await page.goto(server.EMPTY_PAGE);
+  const [response1, response2] = await Promise.all([
+    page.waitForEvent('response', async response => {
+      const text = await response.text();
+      return text.includes('contents of the file');
+    }),
+    page.waitForResponse(async response => {
+      const text = await response.text();
+      return text.includes('bar');
+    }),
+    page.evaluate(() => {
+      fetch('/simple.json').then(r => r.json());
+      fetch('/file-to-upload.txt').then(r => r.text());
+    })
+  ]);
+  expect(response1.url()).toBe(server.PREFIX + '/file-to-upload.txt');
+  expect(response2.url()).toBe(server.PREFIX + '/simple.json');
+});
+
+it('sync predicate should be only called once', async ({page, server}) => {
```

Comment:
why is it specific to sync ?

Reply:
Because async predicate takes time, and we might concurrently call another one if the next event comes quickly enough. For example, responses are coming, and we are fetching the body to find the right response.

Reply:
Makes sense!